commons/src/main/java/org/infinispan/counter/api/CounterManager.java

```diff
@@ -0,0 +1,74 @@
+package org.infinispan.counter.api;
+
+/**
+ * The {@link CounterManager} creates, defines and returns counters.
+ * <p>
+ * It is thread-safe in the way that multiples threads can retrieve/create counters concurrently. If it is the first
+ * time a counter is created, other concurrent threads may block until it is properly initialized.
+ * <p>
+ * A counter can be defined using {@link CounterManager#defineCounter(String, CounterConfiguration)} and {@link
+ * CounterManager#isDefined(String)} returns if the counter is defined or not.
+ * <p>
+ * The counter can be retrieved/created using the {@link CounterManager#getStrongCounter(String)} or {@link
+ * CounterManager#getWeakCounter(String)} to return an (un)bounded strong counter or weak counter. The operation will
+ * fail if the counter is defined with a different type. For example, define a strong counter {@code "test"} and try to
+ * retrieve using the {@code getWeakCounter("test"}.
+ *
+ * @author Pedro Ruivo
+ * @see CounterConfiguration
+ * @since 9.0
+ */
+public interface CounterManager {
+
+   /**
+    * Returns the {@link StrongCounter} with that specific name.
+    * <p>
+    * If the {@link StrongCounter} does not exists, it is created based on the {@link CounterConfiguration}.
+    * <p>
+    * Note that the counter must be defined prior to this method invocation using {@link
+    * CounterManager#defineCounter(String, CounterConfiguration)} or via global configuration. This method only supports
+    * {@link CounterType#BOUNDED_STRONG} and {@link CounterType#UNBOUNDED_STRONG} counters.
+    *
+    * @param name the counter name.
+    * @return the {@link StrongCounter} instance.
+    * @throws org.infinispan.counter.exception.CounterException              if unable to retrieve the counter.
+    * @throws org.infinispan.counter.exception.CounterConfigurationException if the counter configuration is not valid
+    *                                                                        or it does not exists.
+    */
+   StrongCounter getStrongCounter(String name);
```

Comment:
You should specify if the namespace of strong counters and weak counters overlap. In `CounterManagerImpl.counters` you use `Map<String, Object>` for all counters. However, since you have distinct types of keys for weak/strong, if there's a strong key "foo" and you try to retrieve a weak key "foo" on another node, both get created without exception (if I read the code correctly). On the same node it would fail.

Maybe I missed some detail, but I've also missed a test case for this.